### ISSUE TITLE
[RFC][Data][Train] End-to-end observability for the Ray Train + Ray Data training pipeline

### ISSUE THREAD

**Issue by OneSizeFitsQuorum** [ray-project/ray#63911]
# [RFC][Data][Train] End-to-end observability for the Ray Train + Ray Data training pipeline

## Motivation

Ray Train and Ray Data are the two pillars of the Ray training stack. However, their observability systems are developed **independently**, leaving a massive blind spot in the middle — the end-to-end data loading → GPU compute → checkpoint pipeline.

Today, when users ask **"Why is my training slow?"**, they have no way to answer it from Ray's built-in metrics. This RFC proposes a systematic observability initiative to close that gap.

### Ray Data — Rich, Mature Observability

Ray Data exposes **50+ Prometheus metrics** covering data pipeline internals:

| Category | Example Metrics |
|----------|----------------|
| **Operator I/O** | `num_inputs_received`, `bytes_task_outputs_generated`, `rows_task_outputs_generated` |
| **Task scheduling** | `num_tasks_submitted`, `task_scheduling_time`, `block_generation_time` |
| **Memory** | `obj_store_mem_used`, `obj_store_mem_spilled`, `obj_store_mem_freed` |
| **Actors** | `num_alive_actors`, `pool_utilization`, `num_tasks_in_flight` |
| **Resource allocation** | `data_cpu_budget`, `data_gpu_budget`, `data_memory_budget` per operator |
| **Iterator** | `data_iter_total_blocked_seconds`, `data_iter_time_to_first_batch_seconds`, `data_iter_blocks_local/remote` |

All tagged with `dataset` and `operator` identifiers.

### Ray Train — Minimal Observability

Ray Train v2 exposes only **4 metrics**:

| Metric | Description |
|--------|-------------|
| `train_controller_state` | Current state of training controller |
| `train_worker_group_start_total_time_s` | Time to start worker group |
| `train_worker_group_shutdown_total_time_s` | Time to shutdown worker group |
| `train_report_total_blocked_time_s` | Cumulative time to report checkpoint |

Tagged with `ray_train_run_name` and `ray_train_run_id`.

### The Gap — Zero Integration

| Dimension | Ray Data | Ray Train | Connected? |
|-----------|----------|-----------|------------|
| Tags | `dataset`, `operator` | `run_name`, `run_id` | ❌ No correlation |
| Data loading time | Partial (iterator metrics) | ❌ None | ❌ |
| GPU utilization | ❌ None | ❌ None | ❌ |
| Training loop breakdown | N/A | ❌ None | ❌ |
| Prefetch effectiveness | Partial | ❌ None | ❌ |
| Fault tolerance | ❌ None | ❌ None | ❌ |

**Users cannot answer these questions from built-in metrics:**
- Is data loading my bottleneck or is it GPU compute?
- How much time does the GPU spend idle waiting for data?
- What's the end-to-end latency from data read to model update?
- Which operator in my data pipeline is the slowest for this training run?
- How effective is my prefetch buffer?

## Proposed Initiative

We propose a **phased, incremental approach** to build end-to-end observability across the Ray Train + Ray Data training pipeline. Each phase produces an independent, mergeable PR.

### Phase 1: Data Pipeline Observability in `iter_batches` (iter-torch-batches)

**Goal**: Surface CPU-side bottlenecks in the training data loading path.

**Scope** (`python/ray/data/_internal/block_batching/iter_batches.py`):

The `_pipeline()` function has 6 stages, none of which export per-stage Prometheus metrics today:

```
Step 1: Prefetch blocks      → No queue depth or latency metrics
Step 2: Resolve blocks       → No ray.get() overhead metrics
Step 3: blocks_to_batches    → No ShufflingBatcher compaction metrics
Step 4: Format + collate     → No threadpool utilization metrics
Step 5: Finalize             → No GPU transfer latency metrics
Step 6: Restore order        → No reordering overhead metrics
```

**Proposed metrics**:

| Metric | Type | Description |
|--------|------|-------------|
| `ray_data_iter_shuffle_buffer_size_rows` | Gauge | Current shuffle buffer size in rows |
| `ray_data_iter_shuffle_buffer_compactions_total` | Counter | Number of compaction events |
| `ray_data_iter_shuffle_buffer_compaction_seconds` | Histogram | Per-compaction latency |
| `ray_data_iter_prefetch_queue_depth` | Gauge | Number of blocks in prefetch queue |
| `ray_data_iter_format_seconds` | Histogram | Per-batch format time |
| `ray_data_iter_collate_seconds` | Histogram | Per-batch collate time |
| `ray_data_iter_threadpool_active_workers` | Gauge | Active format/collate threads |

**Effort**: ~150-200 lines, low risk, self-contained.

### Phase 2: Training Loop Breakdown Metrics

**Goal**: Expose per-step timing breakdown inside the Ray Train worker loop.

**Scope** (`python/ray/train/v2/`):

| Metric | Type | Description |
|--------|------|-------------|
| `ray_train_step_data_loading_seconds` | Histogram | Time spent loading data per step |
| `ray_train_step_forward_seconds` | Histogram | Forward pass time |
| `ray_train_step_backward_seconds` | Histogram | Backward pass time |
| `ray_train_step_optimizer_seconds` | Histogram | Optimizer step time |
| `ray_train_step_total_seconds` | Histogram | Total step wall time |
| `ray_train_gpu_idle_seconds` | Counter | Cumulative GPU idle time waiting for data |

**Effort**: ~200-300 lines, requires coordination with Train maintainers.

### Phase 3: Cross-Component Correlation

**Goal**: Link Ray Data metrics to Ray Train runs so users can trace end-to-end.

**Scope**: Add shared correlation tags across both systems.

| Change | Component |
|--------|-----------|
| Add `ray_train_run_id` tag to Ray Data iterator metrics when used inside a training worker | Ray Data |
| Add `dataset` tag to Ray Train step metrics | Ray Train |
| Expose a `get_data_loading_stats()` API in Ray Train that returns current iterator metrics | Ray Train |

**Effort**: ~100 lines, but requires design discussion with both teams.

### Phase 4: Prefetch & Async Optimization (Performance)

**Goal**: Based on metrics from Phase 1-3, optimize the actual CPU bottlenecks.

Possible optimizations (guided by real metrics data):
- Adaptive prefetch based on GPU utilization
- Async block resolution (batch `ray.get()`)
- Dynamic threadpool sizing for format/collate

**Effort**: TBD after Phase 1-3 metrics reveal actual bottlenecks.

## Design Principles

1. **Incremental**: Each phase is an independent PR, reviewable and mergeable on its own
2. **Backward compatible**: New metrics are additive; existing metrics unchanged
3. **Opt-in where costly**: Expensive instrumentation behind feature flags
4. **Convention-aligned**: Follow existing Ray Prometheus naming (`ray_data_*`, `ray_train_*`)
5. **Metrics-first**: Build observability before optimization — measure, then improve

## Questions for Maintainers

1. **Scope & priority**: Is this the right direction? Are there other observability gaps you'd prioritize higher?

2. **Metrics namespace**: Should training pipeline metrics live under `ray_data_*`, `ray_train_*`, or a new `ray_training_*` namespace?

3. **Cross-component tagging**: Are there concerns about adding `run_id` tags to Data metrics or `dataset` tags to Train metrics? (Potential cardinality explosion, etc.)

4. **Phase 1 independence**: Can we start with Phase 1 (data pipeline observability in `iter_batches`) as a standalone PR, even if Phase 2-3 aren't committed yet?

5. **Related efforts**: Are there any ongoing observability initiatives in Ray Train v2 that we should align with?

## Related Work

- Issue #42317: Local shuffle buffer throughput regression (closed, but observability gap noted)
- PR #62539: Incremental indexing optimization for `ShufflingBatcher` (merged)
- PR #63744: OTel Prometheus label normalization (merged)

## Environment

- Ray version: master (post 2.55.1)
- Components: Ray Data, Ray Train v2

**Comment by OneSizeFitsQuorum** [comment]
@richardliaw Do you have any thoughts on this issue?

**Comment by abhinavsriva** [comment]
@OneSizeFitsQuorum : This RFC is very aligned with what we have been working on in TraceML.

TraceML has a thin Ray Train integration today: Ray still owns scheduling, worker startup, placement, ranks, and distributed training, while TraceML starts an aggregator actor and a runtime inside each Ray worker.

For Ray Data, because `iter_torch_batches(...)` is not a PyTorch `DataLoader`, we wrap the returned iterator with a small `wrap_dataloader_fetch(...)` helper. That lets us record the `next(iterator)` boundary as input-fetch time and correlate it with step-scoped timings such as H2D transfer, forward, backward, optimizer step, total step wall time, wait proxy, and system GPU utilization.

One limitation of this external approach is that it only sees the iterator boundary. It does not see Ray Data internal stages like prefetch queue depth, block resolution, shuffle-buffer compaction, or format/collate time. That is why Phase 1 seems especially useful: native instrumentation inside `iter_batches` could expose the lower-level cause of input stalls, while external tools can correlate that signal with training-step behavior.

A design question: do maintainers see this primarily as Prometheus metrics inside Ray, or as trace/span-style events that could later be aggregated into metrics?

Happy to share the TraceML Ray example or compare notes if useful. I would also be interested in what native Ray signals would make integrations like ours more useful for Ray Train + Ray Data users.

**Comment by vaishdho1** [comment]
Hi @OneSizeFitsQuorum, this looks like an great addition. I'm interested in contributing to this initiative. Is this open for community contributions?

**Comment by OneSizeFitsQuorum** [comment]
Thanks for the pointers! @abhinavsriva @vaishdho1 

This is my first time learning about TraceML — it looks like a very interesting project, and I’ve starred it.

The motivation behind this RFC comes from my experience using ray data + ray train, where I observed a visibility gap across the end-to-end training path: data loading, iterator-side batching/prefetching, GPU compute, and checkpoint/reporting. It is currently not very straightforward to tell whether a training job is bottlenecked by Ray Data, the Train iterator, GPU compute, or checkpointing.

I’d like to avoid duplicating existing or ongoing efforts. From my recent conversations with folks from Anyscale, it sounds like there may already be some internal RFCs or discussions around this area. We are still discussing how this work should move forward and whether/how parts of it can be aligned with or contributed back to Ray OSS.

My hope is that Ray can eventually provide a clearer set of OSS-level observability primitives for this path — for example, metrics or traces that make it easier to correlate Ray Data execution, Train iteration, GPU utilization, and checkpoint/reporting behavior. I’d be happy to help contribute once we have alignment on the direction, metric boundaries, and what should live in Ray Data, Ray Train, or a shared observability layer.

**Comment by richardliaw** [comment]
Hi @OneSizeFitsQuorum - as you mentioned, we have some internal RFCs here; we'll aim to post something in the next couple days. Let me follow up here when we post that.

**Comment by JasonLi1909** [comment]
Hi @OneSizeFitsQuorum, @abhinavsriva, and @vaishdho1, thanks for the RFC and sharing about TraceML, it’s awesome to see discussions around Ray Data + Train observability and it’s definitely something on our radar. As @richardliaw mentioned, I will put out a doc soon to share more about our internal efforts. @OneSizeFitsQuorum, as you have written the rfc, would you be interested in working on this? 

@ahbi, would be great to hear more about what specific Ray Data iterator metrics or other you’d like to see if not already mentioned in @OneSizeFitsQuorum's rfc.

**Comment by OneSizeFitsQuorum** [comment]
@JasonLi1909 @richardliaw Thanks for the follow-up! I'd be happy to work on this end-to-end — from the iter_batches instrumentation in Phase 1 through the training loop breakdown andcross-component correlation. Looking forward to the internal RFC doc so we can align on the approach before diving in.

**Comment by JasonLi1909** [comment]
Hi @OneSizeFitsQuorum, created an RFC we can start with here: https://github.com/ray-project/ray/issues/64132
